### ISSUE TITLE
LMNN: don't recompute impostors during an SGD batch

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1490]
I debugged the sometimes-hanging `LMNNTest/LMNNLowRankAccuracyBBSGDTest` today.  What I found was that it sometimes hangs because the objective function that is being optimized is actually changing during a single step, because `Evaluate()` and `Gradient()` may be called multiple times, and as a result the impostors may be recalculated and get different results, resulting in a different objective function.

In order to solve this I see a couple options:

1. Only recompute impostors when `Shuffle()` is called.
2. Track which points have had impostors recalculated since the last call to `Shuffle()`, and only recalculate impostors for a point if not.
3. Use some traits class to know at compile time which optimizers may make multiple calls to `Evaluate()` and `Gradient()` during each step.

@manish7294: what do you think, would you be willing to make these changes?  If we don't fix it somehow, techniques like BBSGD won't work with LMNN.

**Comment by rcurtin** [comment]
I also commented out a test in 69ea3599 that suffers from the same issue.

**Comment by manish7294** [comment]
Thanks for looking into this. I have few questions, maybe you can help me out here.

>  What I found was that it sometimes hangs because the objective function that is being optimized is actually changing during a single step, because `Evaluate()` and `Gradient()` may be called multiple times, and as a result the impostors may be recalculated and get different results, resulting in a different objective function.

I don't think I get it completely. Though there are multiple calls to `Evaluate()` and `Gradient()` in a single step but we always update iterate just once during the entire call. So, for one iteration impostors should be recalculated just once. At other times when iterate is not updated bounds should avoid recalculation.

> Only recompute impostors when Shuffle() is called.

I agree this could have been a good option, but I fear this would make learning while using linear scan ineffective. Moreover there is one more reason for avoiding recalulation of impostors during shuffle(), I will write it down in next lines.

> Track which points have had impostors recalculated since the last call to Shuffle(), and only recalculate impostors for a point if not.

In BBSGD, `Shuffle() `is called at very starting of an iteration and the update of iterate takes place after that, so recalculating impostors during `Shuffle()` is unproductive as we won't see any upadates in impostors.

> Use some traits class to know at compile time which optimizers may make multiple calls to Evaluate() and Gradient() during each step.

This could help if we find out a complementary solution for the problem.

Taking from your thoughts, I think problem could be shuffling of those points as well for which impostors recalculation is getting pruned by bounds. And as of now there is no code to handle that in `Shuffle()`

I was trying below code from your `lmnn-tree-rules` for updating target neighbors & impostors, but it didn't seem to work out, though I don't think I see any mistake in there.
```
// Remap target neighbors to new indices.
  arma::Mat<size_t> newNeighbors(targetNeighbors.n_rows,
                                 targetNeighbors.n_cols);
  for (size_t i = 0; i < targetNeighbors.n_cols; ++i)
  {
    const size_t newCol = ordering[i];
    for (size_t j = 0; j < targetNeighbors.n_rows; ++j)
      newNeighbors(j, newCol) = ordering(targetNeighbors(j, i));
  }
  targetNeighbors = std::move(newNeighbors);
```

**Comment by rcurtin** [comment]
> I don't think I get it completely. Though there are multiple calls to Evaluate() and Gradient() in a single step but we always update iterate just once during the entire call. So, for one iteration impostors should be recalculated just once. At other times when iterate is not updated bounds should avoid recalculation.

Consider if my optimizer works like this:

```
Evaluate(coordinates);
Gradient(coordinates, gradient);
for (size_t i = 0; i < 10; ++i)
{
  stepSize += someFunction(Evaluate(coordinates + i));
}

coordinates = stepSize * gradient;
```

In this case, the impostors will be recalculated at a different transformation matrix during that loop.  For BBSGD, this is basically exactly what happens---we compute the gradient at the current transformation matrix, then try a bunch of different steps in a line search to see which is best.

But every time we call `Evaluate()`, we recompute impostors---which might be different for different coordinates.  For BBSGD, this can cause the line search to fail, because BBSGD (and most optimizers) expect that the function being optimized does not change during a single step of the optimization.  (And when I say change, recomputing impostors and getting different results counts for that.)

Does the problem make more sense now?  If so then we can talk about solution ideas in more details.

**Comment by manish7294** [comment]
Thanks! It's clear now. Now, I can see that these multiple calls between each iteration can cause impostors to change drastically leading to uneven updates in coordinates.

**Comment by rcurtin** [comment]
Ok, great, now let's talk about how to solve it. :)

>> Only recompute impostors when Shuffle() is called.
>
> I agree this could have been a good option, but I fear this would make learning while using linear scan ineffective. Moreover there is one more reason for avoiding recalulation of impostors during shuffle(), I will write it down in next lines.

Hmm, but in the case of linear scan, since we call a different overload of `Evaluate()` or `Gradient()` or `EvaluateWithGradient()`, we can still compute impostors in those cases.

>> Track which points have had impostors recalculated since the last call to Shuffle(), and only recalculate impostors for a point if not.
>
> In BBSGD, Shuffle()is called at very starting of an iteration and the update of iterate takes place after that, so recalculating impostors during Shuffle() is unproductive as we won't see any upadates in impostors.

Well, but an update would have happened at the end of the previous iteration, so I think there is no issue there.  A problem with this approach is that there's a lot of extra overhead in tracking which points have had impostors recalculated and which haven't.

A traits class is possible but I think it would be a decent amount of code and probably more work than the other approaches.

So, with all that said, which of these three options would you prefer?

**Comment by manish7294** [comment]
Hmm, then I think the first one will be the optimal choice for taking time and the work needed to be done into consideration :)

Just need to clarify this part before starting work - 
>Hmm, but in the case of linear scan, since we call a different overload of Evaluate() or Gradient() or EvaluateWithGradient(), we can still compute impostors in those cases. 

By different overload of Evaluate() or Gradient() or EvaluateWithGradient() do you mean the one which works on full dataset. Can you clarify this a bit?

And basically we will be removing the impostors recalculation code from Evaluate() and Gradient() related to bbsgd (ones with batch overload) and will have recalculation code in  shuffle(), right?

**Comment by rcurtin** [comment]
> By different overload of Evaluate() or Gradient() or EvaluateWithGradient() do you mean the one which works on full dataset. Can you clarify this a bit?

Right, exactly.

> And basically we will be removing the impostors recalculation code from Evaluate() and Gradient() related to bbsgd (ones with batch overload) and will have recalculation code in shuffle(), right?

Yeah, I think this is the right way to do it.  This may also accelerate some of the amsgrad/SGD runs too.

**Comment by manish7294** [comment]
Just one more question, I am not seeing anything like Evaluate(iterate) in bigbatch_sgd code, so don't you think it will just be calling `Evaluate(iterate, currentFunction, effectiveBatchSize); `even in the case of linear_scan. 
for linear scan we only have-
```
if (shuffle) // Determine order of visitation.
        f.Shuffle();
```

**Comment by rcurtin** [comment]
You're right, I was just making up a code example there.  Sorry for the incorrectness.

**Comment by rcurtin** [comment]
Also, sorry, I was mixing up linear scan and the full-batch L-BFGS type optimizers!  Hmm.  That does make this more difficult.  Let me think for a little bit about how to work around that situation and make sure impostors are still recalculated at the beginning of an epoch, and maybe you can think about it too and we can see what we come up with.

**Comment by manish7294** [comment]
Sure. This isn't related to this issue, but can you pick and merge the changes made in #1472 related to LMNN accuracy tests (lmnn_test.cpp) except `UpdateTreeTest` for the time being.

**Comment by rcurtin** [comment]
If you can make those into a separate PR, I'm happy to look through and approve it.

**Comment by rcurtin** [comment]
All I can think of is the idea of holding a variable `linearScan` in `LMNNFunction`, and using that in `Shuffle()` to decide whether to actually do anything (and computing impostors either way).  Then we would have to add documentation stating that linear scan should be disabled when an LMNNFunction is being optimized, otherwise impostor recalculation will not happen and that can cause worse results.

Can you come up with anything better?  That's the only way I can see that's not a massive amount of refactoring work.

**Comment by manish7294** [comment]
That sounds fair. But shouldn't we be using `linearScan` to decide whether to do recalculation in Evaluate() and Gradient() instead of using it in shuffle(). As when we never make call to shuffle() when linear_scan is used. That way we won't be removing recalculation code from Evaluate() and Gradient(), just will have an `if (linearScan)` condition over it.

Sorry for asking it here, since the GSoC final evaluations are started, they have asked to make a blog or something like that. So, can you please give me some tips on how it should be and work upto which date should be included.

**Comment by rcurtin** [comment]
I'll send an email about the final evaluations in just a moment. :)

I don't understand how your idea would work---the whole issue is that with some optimizers, `Evaluate()` or `Gradient()` may be called arbitrary numbers of times before a step is taken.  But we need to make sure that the impostors are the same during this process.  So even if we know what `linearScan` is, how could we ensure that we don't recalculate impostors when we should?

**Comment by manish7294** [comment]
Ah!  Right, Sorry I missed the main point while writing that. But then in linearScan there is no call to shuffle, so how we are going to recalculate impostors, when we don't have any code in Evaluate() and we can't access shuffle at that time?

**Comment by rcurtin** [comment]
Yes, we would disable linear scan in the optimizers and only set it in `LMNNFunction`.  So any optimizer would believe it was not doing a linear scan and call `Shuffle()` at the start of every epoch.

**Comment by rcurtin** [comment]
I thought also, if we implement this by only computing impostors once at the beginning of each epoch, it will make #1446 a lot easier also.

**Comment by manish7294** [comment]
I was working on this and observed a few obstacles. 
For recomputing at the start of every epoch will require us to access `begin` of each batch inside shuffle().
At the same time, we need access to updated iterate in the shuffle() as well. But the problem is shuffle() is called before iterate update call and there doesn't seem any easy way to access it.

**Comment by rcurtin** [comment]
Hi Manish:

> For recomputing at the start of every epoch will require us to access begin of each batch inside shuffle().

I don't understand why this would be true.

> At the same time, we need access to updated iterate in the shuffle() as well. But the problem is shuffle() is called before iterate update call and there doesn't seem any easy way to access it.

Why don't you do what you did in `Precalculate()` and use a boolean variable indicating whether the impostors need to be recalculated at the start of a call to `Evaluate()`, `Gradient()`, or `EvaluateWithGradient()`?

**Comment by rcurtin** [comment]
I've already answered this question:

> All I can think of is the idea of holding a variable linearScan in LMNNFunction, and using that in Shuffle() to decide whether to actually do anything (and computing impostors either way). Then we would have to add documentation stating that linear scan should be disabled when an LMNNFunction is being optimized, otherwise impostor recalculation will not happen and that can cause worse results.
